server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java

```diff
@@ -150,19 +153,27 @@ final class DefaultSearchContext extends SearchContext {
         this.indexShard = readerContext.indexShard();
 
         Engine.Searcher engineSearcher = readerContext.acquireSearcher("search");
-        this.searcher = new ContextIndexSearcher(
-            engineSearcher.getIndexReader(),
-            engineSearcher.getSimilarity(),
-            engineSearcher.getQueryCache(),
-            engineSearcher.getQueryCachingPolicy(),
-            minimumDocsPerSlice,
-            lowLevelCancellation,
-            // TODO not set the for now, this needs a special thread pool and can be enabled after its introduction
-            // parallelize
-            // ? (EsThreadPoolExecutor) this.indexService.getThreadPool().executor(ThreadPool.Names.CONCURRENT_COLLECTION_TBD)
-            // : null,
-            null
-        );
+        if (executor == null) {
+            this.searcher = new ContextIndexSearcher(
+                engineSearcher.getIndexReader(),
+                engineSearcher.getSimilarity(),
+                engineSearcher.getQueryCache(),
+                engineSearcher.getQueryCachingPolicy(),
+                lowLevelCancellation
+            );
+        } else {
+            this.searcher = new ContextIndexSearcher(
+                engineSearcher.getIndexReader(),
+                engineSearcher.getSimilarity(),
+                engineSearcher.getQueryCache(),
+                engineSearcher.getQueryCachingPolicy(),
+                lowLevelCancellation,
+                executor,
+                minimumDocsPerSlice,
+                // TODO is there a better way that does not require casting the executor?
+                forceSequentialCollection ? 1 : ((ThreadPoolExecutor) executor).getMaximumPoolSize()
```

Comment:
this cast is not great. We had it before, and I have made it effective only when concurrency is enabled. There was SnapshotResiliencyTests that injects a different executor and was causing a class cast exception here, to I went for disabling concurrency in that test entirely via the cluster setting.

We could also expose the max pool size (searchOrGetThreadPoolSize) in some way in ThreadPool instead of retrieving it from the thread pool executor, but I am not too sure that that is better.

Reply:
I think it might be a bit nicer to handle this case locally rather than changing settings in `SnapshotResiliencyTests` (which might never get reverted) - something like `executor instanceof ThreadPoolExecutor threadPoolExecutor ? threadPoolExecutor.getMaximumPoolSize() : 1` perhaps?

Reply:
Ok, I am not sure which is best: disable concurrency in the test that has the funky thread pool or silently disable it in prod code whenever you have a funky thread pool. The latter seems scary to me, is it a situation that could happen outside of tests?

Reply:
If you're using an executor that isn't from a threadpool then its max pool size is effectively `1` anyway (it's a thread on its own). I think I'd rather fall back to some sensible behaviour instead of failing with a `ClassCastException` if ever we encounter an inappropriate executor, but I'm also pretty sure we wouldn't see an inappropriate executor in prod code today.

Perhaps slightly neater would be to introduce a new interface with a method for this specific case, and apply it to both `EsThreadPoolExecutor` and the ones used in tests. I can see that being more generally useful, there are several other places where we try and size some pool of workers according to the size of the underlying threadpool.

Reply:
I went for your initial suggestion, thanks for the feedback.